### DISCUSSION TITLE
Any way to use math expression in binding

### DISCUSSION THREAD

**Discussion by ds1709** [AvaloniaUI/Avalonia#16812] [Q&A]
In version `11.0.*` I implemented my own IBinding, using math expression. For example, for class
```csharp
public class Foo
{
    public int A { get; set; }
    public int B { get; set; }
    public int C { get; set; }
}
```
I was able to write something like this
```xml
<Border>
    <Border.BorderThickness>
        <MathExpressionBinding Expression="A + B + C" Converter="{StaticResource DecimalToThicknessConverter}" />
    </Border.BorderThickness>
</Border>
```
This binding tracks properties `A`, `B` and `C`, and updates `BorderThickness` when any of them changes.

In version `11.1.*` custom `IBinding`-s are disallowed. So, is there any way to achive such result in version `11.1.*`?

Using of `MultiBinding` is bulky and requires to write `MultiConverters`. So, in case with `BorderThickness`, I need a simple converter when bind to single value, and a multiconverter when bind to multiple values. Also using of multibindings significantly reduces readability because instead of `A`, `B` and `C` I must use something like `arg[0]`, `arg[1]` or `arg[2]`.

**Comment by MikeCodesDotNET** [comment]
This seems like a duplicate of your [previous question](https://github.com/AvaloniaUI/Avalonia/issues/16623)

**Reply by ds1709** [reply]
Not really. Previous question was about broken backward compatibility. In this question I just want to now, is there any workaround when you migrate from `11.0.*` to `11.1.*`.

**Comment by thevortexcloud** [comment]
I am pretty sure you can still inherit from `Binding`/`BindingBase`, just not from `IBinding` if you _really_ need to do this as a custom binding

https://github.com/AvaloniaUI/Avalonia/blob/06ece8da8813119b0feb0d35de397dd031157095/src/Markup/Avalonia.Markup/Data/BindingBase.cs#L12

https://github.com/AvaloniaUI/Avalonia/blob/06ece8da8813119b0feb0d35de397dd031157095/src/Markup/Avalonia.Markup/Data/Binding.cs#L19

Otherwise you can probably do something with a custom markup extension that accepts a regular binding as an input. Although getting the actual value from a binding is not always straight forwards.

**Reply by ds1709** [reply]
> I am pretty sure you can still inherit from `Binding`/`BindingBase`.

No, you can't. In version `11.1.*` it contains `private protected abstract` method, that can't be implemented anywhere outside their assembly.

**Comment by stevemonaco** [comment]
You can instead create a `MarkupExtension` that returns an `IBinding` instance. One problem is that "A + B + C" requires reflection to resolve the properties. Maybe some magic with subscribing to `PropertyChanged` and filtering if they're always local properties, otherwise you would have to bind arguments and index them in the expression.

You could also expose the mathematical operation as an `IObservable<T>` on the View or ViewModel to bind to, eliminating the particular problem from the XAML side.

Some guidance/support on the suggested approach for this from the team would be really useful.
